core/service/src/engine/threshold/service/resharer.rs

```diff
@@ -370,3 +591,214 @@ impl<PubS: Storage + Send + Sync + 'static, PrivS: Storage + Send + Sync + 'stat
+            RamStorage::new(),
+            RamStorage::new(),
+            None,
+            HashMap::new(),
+            HashMap::new(),
+        );
+
+        let context_id = ContextId::new_random(&mut rng);
+        let context_info = ContextInfo {
+            mpc_nodes: vec![NodeInfo {
+                mpc_identity: "Node1".to_string(),
+                party_id: 1,
+                verification_key: None,
+                external_url: "http://localhost:12345".to_string(),
+                ca_cert: None,
+                // the storage url does not matter as we're using the mock
+                public_storage_url: "https://zama-zws-dev-tkms-b6q87.s3.eu-west-1.amazonaws.com/"
+                    .to_string(),
+                extra_verification_keys: vec![],
+            }],
+            context_id,
+            software_version: SoftwareVersion {
+                major: 0,
+                minor: 1,
+                patch: 0,
+                tag: None,
+            },
+            threshold: 0,
+            pcr_values: vec![],
+        };
+
+        crypto_storage
+            .inner
+            .write_context_info(&context_id, &context_info)
+            .await
+            .unwrap();
+
+        let ro_storage_getter = DummyRealReadOnlyS3StorageGetter { ram_storage };
+        {
+            // sunshine
+            // use the dummy s3 storage to fetch the keys from ram storage
+            let _keyset = fetch_public_materials_from_peers::<_, _, _, DummyS3StorageReadOnly>(
+                &crypto_storage,
+                key_id,
+                &context_id,
+                &key_digests,
+                &ro_storage_getter,
+            )
+            .await
+            .unwrap();
+        }
+        {
+            // negative test
+            // use wrong digests to trigger error
+            let wrong_key_digests: HashMap<PubDataType, Vec<u8>> = HashMap::from_iter([
+                (PubDataType::ServerKey, vec![0, 1, 2, 4]),
+                (PubDataType::PublicKey, vec![3, 4, 5, 6]),
+            ]);
+            let _keyset = fetch_public_materials_from_peers::<_, _, _, DummyS3StorageReadOnly>(
+                &crypto_storage,
+                key_id,
+                &context_id,
+                &wrong_key_digests,
+                &ro_storage_getter,
+            )
+            .await
+            .unwrap_err();
+        }
+        {
+            // negative test
+            // use empty storage to trigger error
+            let ro_storage_getter = DummyRealReadOnlyS3StorageGetter {
+                ram_storage: RamStorage::new(),
+            };
+            let _keyset = fetch_public_materials_from_peers::<_, _, _, DummyS3StorageReadOnly>(
+                &crypto_storage,
+                key_id,
+                &context_id,
+                &key_digests,
+                &ro_storage_getter,
+            )
+            .await
+            .unwrap_err();
```

Comment:
Could you either use custom errors or maybe do a traced test and check that it is indeed the expected error that is logged?

Reply:
added test to check for error messages in 32c2f1785e6d0b62fb482b41d104a0d986c82d2e